Intersections: Consistent behavior of Circle_2 and Sphere_3

soesau opened the pull request:
## Summary of Changes

Circle_2 and Sphere_3 intersections are not fully consistent. The Circle_2 - Segment_2 intersection does consider inclusion as intersection. The same is true for Sphere_3 and Iso_cuboid_3.

This PR changes the behavior of those two intersection methods and adds specific intersection methods with inclusion for Circle_2 and Iso_rectangle_2 to AABB_traits_2 with static filter. The same will be done for Sphere_3, Iso_cuboid_3 and AABB_traits_3.

## Release Management

* Affected package(s): AABB_tree, Intersection_2 Intersection_3
* Issue(s) solved (if any): fix #8907, fix #8891, fix #6404

soesau committed:
adding filtered and static filtered AABB_traits_2

soesau committed:
Circle_2 - Segment_2 intersection does now consider inclusion as non intersecting

soesau committed:
updated AABB_tree dependencies

soesau committed:
moved filtered AABB_traits into own files
added filtered traits for 3d AABB tree
changed behavior of Sphere_3 Iso_cuboid_3 intersection

soesau committed:
fixed point type

soesau committed:
updated dependencies

soesau commented:
documentation still needs to be updated, but can be tested already

soesau committed:
fixed warning

soesau committed:
removing sphere/box intersection test from Orthtree

soesau committed:
fixed types in AABB_traits_3::Compare_distance

soesau committed:
fixed types in AABB_traits_3::Compare_distance & Custom_traits_Hausdorff

soesau committed:
make filtered kernel types public

soesau committed:
removed typename

soesau committed:
bugfix

soesau committed:
added missing operators in derived class

soesau committed:
fixing types for filtered kernels

soesau committed:
simplifying generic iso_cuboid - sphere intersection tests

soesau committed:
resetting number of generic sphere/iso_cuboid_3 intersection tests

sloriot commented:
Looks in CGAL-6.2-Ic-30 but the PR is still under review.

soesau committed:
removing AABB_filtered_traits_3.h and AABB_statically_filtered_traits_3.h

soesau committed:
removing AABB_filtered_traits_2.h and AABB_statically_filtered_traits_2.h

soesau committed:
fixing name collision

soesau committed:
adding static filter for Sphere_3 Iso_cuboid_3 intersection

soesau committed:
adding breaking changes for intersections to CHANGES.md
[skip ci]

soesau committed:
setting Has_static_filters for Epeck to true
fixing static filter for Sphere_3 Iso_cuboid_3 intersection

soesau committed:
bugfix

soesau committed:
fixed some documentation

sloriot commented on Filtered_kernel/test/Filtered_kernel/Static_filters.cpp:
```diff
@@ -309,20 +309,6 @@ int main(int argc, char **argv)
   assert( !CGAL::Epick::Has_static_filters);
   #endif
 
-  #ifdef CGAL_DONT_USE_LAZY_KERNEL
```
why removing this?

soesau commented on Orthtree/include/CGAL/Orthtree.h:
```diff
@@ -1276,15 +1276,6 @@ class Orthtree {
     return recursive_descendant(child(node, i), remaining_indices...);
   }
 
-  bool do_intersect(Node_index n, const Sphere& sphere) const {
```
do_intersect(Node_index n, const Sphere& sphere) was a private method that was used in the recursive nearest neighbors method. However, the nodes are sorted with respect to the distance. Thus, most of the work for calculating an intersection is already done. The new version benchmarked a bit faster in the tests I did.

The change for intersected_nodes makes sense. I can add it.

soesau replied:
I added intersected_nodes using a functor. However, I don't have a solution for d dimensional so the method without provided functor is broken in case of dimension>3

soesau committed:
added intersected_nodes with functor

soesau committed:
changed default functor for intersected_nodes

soesau committed:
bugfix

sloriot committed:
copy/paste error

sloriot commented on Orthtree/include/CGAL/Orthtree.h:
```diff
@@ -782,7 +782,28 @@ class Orthtree {
   template <typename Query, typename OutputIterator>
   OutputIterator intersected_nodes(const Query& query, OutputIterator output) const {
-    return intersected_nodes_recursive(query, root(), output);
+    return intersected_nodes_recursive(query, root(), output, [](const Query& query, const typename Traits::Bbox_d &box) -> bool {return CGAL::do_intersect(query, box);});
```
I believe we should provide an API for an intersection with a ball, as it is a common operation, say `intersected_nodes(query, center, radius, output)`

sloriot commented on Intersections_2/include/CGAL/Intersections_2/Circle_2_Segment_2.h:
```diff
@@ -13,32 +13,109 @@
+  else if (ltgt <= sqrad)
+    return true;
+
+  return sq_dist(c, s) <= sqrad;
```
here we already know that the orthogonal projection of the center is within src and tgt so we can use the line (to not redo the tests)

sloriot commented on Intersections_2/include/CGAL/Intersections_2/Circle_2_Segment_2.h:
```diff
@@ -13,32 +13,109 @@
+  else if (ltgt <= sqrad * ltgtw)
+    return true;
+
+  return sq_dist(c, s) <= sqrad;
```
same comment as below

sloriot commented on Installation/CHANGES.md:
```diff
@@ -1,5 +1,8 @@
 
+### [2D and 3D Linear Geometry Kernel
+- **Breaking change**: Circle_2/Segment_2, Sphere_3/Bbox_3, Sphere_3/Iso_cuboid_3 now do not consider inclusion as intersection. This behavior is consistent with other intersections involving Circle_2 and Sphere_3.
+
```
shall Octree/quadtree be mentioned?

sloriot committed:
update bounds with Filtered Predicate Generator

soesau committed:
optimization
removal of unnecessary homogeneous implementation

soesau committed:
added Circle_2::has_on_bounded_side(Segment_2)

soesau committed:
added Sphere_3::has_on_bounded_side(Iso_cuboid_3)

soesau committed:
ci fix

soesau committed:
ci fix

soesau committed:
moved on_bounded_side check into Has_on_bounded_side_2/3
added static filter for Sphere_3 and Iso_cuboid_3

soesau committed:
adding static filter for Has_on_bounded_side_2 for Circle_2 & Segment_2
bugfix of Sphere_3 tests

soesau committed:
added changes in concepts
added how to reproduce former behavior to CHANGES.md

soesau committed:
Merge branch 'main' into Intersection-Circle_Sphere_consistency-GF

soesau committed:
added Boolean type to AABBGeomTraits_2 and AABBGeomTraits_3
[skip ci]

soesau committed:
added Has_on_bounded_side_2::operator() for Circle_2 and Iso_rectangle_2 including static filter
added intersected_nodes for disk/ball to Orhtree

MaelRL commented on Kernel_23/doc/Kernel_23/Concepts/FunctionObjectConcepts.h:
```diff
@@ -8091,10 +8092,24 @@ class HasOnBoundedSide_2 {
 
+  /*!
+    returns true iff `s` lies on the bounded side of `c`.
+    The endpoints of `s` are allowed to be on the boundary of `c`.
```
What I mean is that you have an inconsistency between existing overloads and new overloads within the functor `Has_on_bounded_side` because you can have:
```cpp
has_on_bounded_side(seg) == true
```
and
```cpp
has_on_bounded_side(seg.source()) == false
```

If you only need knowledge of a `do_intersect()` with the n-ball, you could in fact use "Has_on_unbounded" instead: Has_on_unbounded(circle, segment) == false <==> the segment intersects the ball

soesau committed:
removed Boolean type (replaced by auto)
fixed concepts (review)
changed Has_on_bounded_side_2/3 to require Segment_2, Iso_rectangle_2 and Iso_cuboid_3 to be fully inside the Circle_2/Sphere_3

soesau committed:
added Has_on_unbounded_side_2/3 operators for Circle_2/Segment_2, Circle_2/Iso_rectangle_2 and Sphere_3/Iso_cuboid_3

soesau committed:
wrong point type

soesau committed:
updated CHANGES.md
[ci skip]

soesau committed:
updated CHANGES.md
[ci skip]

soesau committed:
fix Has_on_unbounded_side_2/3 for Circle_2 and Sphere_3

sloriot committed:
fix has_on_unbounded_side for circle vs segment

+ remove newly added Has_on_bounded_side overloads

sloriot commented:
I got rid of all Has_on_bounded_side overloads added (that we can re-add in another PR using code from c04efc3a516ffb01e81ed7cf586b4ce836c53eed). I also fixed the Has_on_unbounded_side as reported in my minimal example above, that I also added as a test. I removed the corresponding static filter. We can readd it if needed.
